tests/unittests/ExecutorTest.cpp

```diff
@@ -0,0 +1,501 @@
+  //         /    \
+  //        v     v
+  //     delta   eps
+  //
+  ExecutorTestDAG dag;
+  DAGNode *alpha = dag.addNode("alpha", testDeviceId, /*parents=*/{},
+                               /*inputs=*/{alphaIn.get()},
+                               /*outputs=*/{alphaOut.get()});
+  DAGNode *beta = dag.addNode("beta", testDeviceId, /*parents=*/{},
+                              /*inputs=*/{betaIn.get()},
+                              /*outputs=*/{betaOut.get()});
+  DAGNode *gamma = dag.addNode("gamma", testDeviceId, /*parents=*/{alpha, beta},
+                               /*inputs=*/{alphaOut.get(), betaOut.get()},
+                               /*outputs=*/{deltaIn.get(), epsIn.get()});
+  dag.addNode("delta", testDeviceId, /*parents=*/{gamma},
+              /*inputs=*/{deltaIn.get()},
+              /*outputs=*/{deltaOut.get()});
+  dag.addNode("eps", testDeviceId, /*parents=*/{gamma},
+              /*inputs=*/{epsIn.get()},
+              /*outputs=*/{epsOut.get()});
+
+  // Create the input Context to pass to Executor::run().
+  auto inputContext = std::make_unique<Context>(
+      cloneContext(ctx_, /*whitelist=*/{alphaIn.get(), betaIn.get()}));
+
+  // Create input/output Contexts for all DAGNodes. The input Contexts are
+  // expected to be passed to TestDeviceManager::run() when the DAGNodes are
+  // run and the output Contexts should be returned by TestDeviceManager
+  // (via callback) when the corresponding DAGNodes are run.
+  auto alphaInputContext = std::make_unique<Context>(
+      cloneContext(ctx_, /*whitelist=*/{alphaIn.get()}));
+  auto alphaOutputContext = std::make_unique<Context>(
+      cloneContext(ctx_, /*whitelist=*/{alphaOut.get()}));
+  auto betaInputContext = std::make_unique<Context>(
+      cloneContext(ctx_, /*whitelist=*/{betaIn.get()}));
+  auto betaOutputContext = std::make_unique<Context>(
+      cloneContext(ctx_, /*whitelist=*/{betaOut.get()}));
+  auto gammaInputContext = std::make_unique<Context>(
+      cloneContext(ctx_, /*whitelist=*/{alphaOut.get(), betaOut.get()}));
+  auto gammaOutputContext = std::make_unique<Context>(
+      cloneContext(ctx_, /*whitelist=*/{deltaIn.get(), epsIn.get()}));
+  auto deltaInputContext = std::make_unique<Context>(
+      cloneContext(ctx_, /*whitelist=*/{deltaIn.get()}));
+  auto deltaOutputContext = std::make_unique<Context>(
+      cloneContext(ctx_, /*whitelist=*/{deltaOut.get()}));
+  auto epsInputContext = std::make_unique<Context>(
+      cloneContext(ctx_, /*whitelist=*/{epsIn.get()}));
+  auto epsOutputContext = std::make_unique<Context>(
+      cloneContext(ctx_, /*whitelist=*/{epsOut.get()}));
+
+  // Create the Context that Executor::run() is expected to return.
+  auto refResultContext =
+      cloneContext(ctx_, /*whitelist=*/{deltaOut.get(), epsOut.get()});
+
+  // Register the expected input Contexts and desired output Contexts for
+  // all nodes with the TestDeviceManager.
+  deviceManager->registerResult("alpha", 0, ResultCode::Executed,
```

Comment:
Similar issue here. I thought about folding this into `ExecutorTestDAG` but I think that class will be more useful if it is focused only on building valid DAGs.